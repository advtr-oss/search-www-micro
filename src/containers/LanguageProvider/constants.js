/*
 *
 * LanguageProvider constants
 *
 */

export const CHANGE_LOCALE = 'app/LanguageToggle/CHANGE_LOCALE';

export const LOCALE_KEY = 'locale'
