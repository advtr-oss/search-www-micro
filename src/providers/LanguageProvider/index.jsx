/**
 * @advtr/tidy tries to be un-opinionated, so for this we need to import `LanguageProvider` and wrap it for redux
 * */

import { connect } from 'react-redux'
import { createSelector } from 'reselect'
import { LanguageProvider } from '@advtr/tidy'

import { makeSelectLocale } from './selectors'

const mapStateToProps = createSelector(
  makeSelectLocale(),
  locale => ({
    locale
  })
)

function mapDispatchToProps (dispatch) {
  return {
    dispatch
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LanguageProvider)
