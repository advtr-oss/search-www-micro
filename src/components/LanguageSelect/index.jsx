/**
 * LanguageSelect
 *
 * This is a basic select that allows for the changing of a locale by the user
 *
 * @note: `@advtr/tidy` tries to be un-opinionated, so for this we need to import `LanguageSelect` and wrap it for redux
 * */

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { createSelector } from 'reselect'

import { LanguageSelect as ADVLanguageSelect } from '@advtr/tidy'

import messages from './messages'
import { appLocales } from '../../i18n'
import { changeLocale } from '../../providers/LanguageProvider/actions'
import { makeSelectLocale } from '../../providers/LanguageProvider/selectors'

function LanguageSelect (props) {
  return <ADVLanguageSelect locale={props.locale} locales={appLocales} messages={messages} onLocaleToggle={props.onLocaleToggle} />
}

LanguageSelect.propTypes = {
  onLocaleToggle: PropTypes.func,
  locale: PropTypes.string
}

const mapStateToProps = createSelector(
  makeSelectLocale(),
  locale => ({
    locale
  })
)

export function mapDispatchToProps (dispatch) {
  return {
    onLocaleToggle: evt => dispatch(changeLocale(evt.target.value)),
    dispatch
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LanguageSelect)
