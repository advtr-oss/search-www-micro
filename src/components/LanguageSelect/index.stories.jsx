import React from 'react'

import ADVLanguageSelect from '../LanguageSelect'
import LanguageProvider from '../../providers/LanguageProvider'

import { translationMessages } from '../../i18n'

export default {
  title: 'Select/Language',
  component: ADVLanguageSelect
}

const Template = () => {
  return (
    <LanguageProvider messages={translationMessages}>
      <ADVLanguageSelect />
    </LanguageProvider>
  )
}

export const LanguageSelect = Template.bind({})
LanguageSelect.args = {}
