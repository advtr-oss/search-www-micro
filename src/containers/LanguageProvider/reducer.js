/*
 *
 * LanguageProvider reducer
 *
 * Now actively looks for the users language
 *
 * @note: Not the best but will work
 */

import produce from 'immer'
import get from 'lodash.get'
import split from 'lodash.split'
import includes from 'lodash.includes'

import { CHANGE_LOCALE, LOCALE_KEY } from './constants'
import { DEFAULT_LOCALE, appLocales } from '../../i18n'

// Detect user language.
const userLanguage =
  window.localStorage.getItem(LOCALE_KEY) ||
  window.navigator.language ||
  window.navigator.userLanguage

let foundLanguage = includes(appLocales, userLanguage) && userLanguage

if (!foundLanguage) {
  // Split user language in a correct format.
  const userLanguageShort = get(split(userLanguage, '-'), '0')

  // Check that the language is included in the admin configuration.
  foundLanguage = includes(appLocales, userLanguageShort) && userLanguageShort
}

export const initialState = {
  locale: foundLanguage || DEFAULT_LOCALE
}

/* eslint-disable default-case, no-param-reassign */
const languageProviderReducer = (state = initialState, action) =>
  produce(state, draft => {
    switch (action.type) {
      case CHANGE_LOCALE:
        window.localStorage && window.localStorage.setItem(LOCALE_KEY, action.locale)
        draft.locale = action.locale
        break
    }
  })

export default languageProviderReducer
