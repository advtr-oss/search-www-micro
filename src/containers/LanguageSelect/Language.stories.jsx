import React from 'react'

import ADVLanguageSelect from "../LanguageSelect"
import LanguageProvider from "../LanguageProvider";

import { translationMessages } from '../../i18n'

export default {
  title: 'UI/LanguageSelect',
  component: ADVLanguageSelect
}

const Template = () => {
  return (
    <LanguageProvider messages={translationMessages}>
      <ADVLanguageSelect />
    </LanguageProvider>
  )
}

export const LanguageSelect = Template.bind({})
LanguageSelect.args = {}
