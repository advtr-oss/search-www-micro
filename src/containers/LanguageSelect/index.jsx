/*
 *
 * LanguageSelect
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { createSelector } from 'reselect'

import { LanguageSelect as ADVLanguageSelect } from '@advtr/tidy'

import messages from './messages'
import { appLocales } from '../../i18n'
import { changeLocale } from '../LanguageProvider/actions'
import { makeSelectLocale } from '../LanguageProvider/selectors'

function LanguageSelect (props) {
  return <ADVLanguageSelect locale={props.locale} locales={appLocales} messages={messages} onLocaleToggle={props.onLocaleToggle} />
}

LanguageSelect.propTypes = {
  onLocaleToggle: PropTypes.func,
  locale: PropTypes.string
}

const mapStateToProps = createSelector(
  makeSelectLocale(),
  locale => ({
    locale
  })
)

export function mapDispatchToProps (dispatch) {
  return {
    onLocaleToggle: evt => dispatch(changeLocale(evt.target.value)),
    dispatch
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LanguageSelect)
