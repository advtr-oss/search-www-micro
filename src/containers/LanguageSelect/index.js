/*
 *
 * LanguageSelect
 *
 */

import React from 'react'
import PropTypes from 'prop-types'
import { connect } from 'react-redux'
import { createSelector } from 'reselect'

import { Select } from '../../components/LanguageSelect'
import Wrapper from './Wrapper'
import messages from './messages'
import { appLocales } from '../../i18n'
import { changeLocale } from '../LanguageProvider/actions'
import { makeSelectLocale } from '../LanguageProvider/selectors'

export function LanguageSelect (props) {
  return (
    <Wrapper>
      <Select
        value={props.locale}
        values={appLocales}
        messages={messages}
        onToggle={props.onLocaleToggle}
      />
    </Wrapper>
  )
}

LanguageSelect.propTypes = {
  onLocaleToggle: PropTypes.func,
  locale: PropTypes.string
}

const mapStateToProps = createSelector(
  makeSelectLocale(),
  locale => ({
    locale
  })
)

export function mapDispatchToProps (dispatch) {
  return {
    onLocaleToggle: evt => dispatch(changeLocale(evt.target.value)),
    dispatch
  }
}

export default connect(
  mapStateToProps,
  mapDispatchToProps
)(LanguageSelect)
