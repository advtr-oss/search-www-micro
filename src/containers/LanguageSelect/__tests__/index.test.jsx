/**
 * @jest-environment jsdom
 */

import React from 'react'
import { Provider } from 'react-redux'
import { render } from '@testing-library/react'
import user from '@testing-library/user-event'
import { GlobalStyle, ThemeProvider } from '@advtr/tidy'

import LanguageSelect from "../index"
import LanguageProvider from "../../LanguageProvider";

import { translationMessages, DEFAULT_LOCALE } from '../../../i18n'
import configureStore from '../../../configureStore'

const initialState = {}
const store = configureStore(initialState, history)

// Add theme provider here too, will look at moving this out and adding a generic one
const redux = (ui, opts) => render((
  <Provider store={store}>
    <ThemeProvider>
      <React.Fragment>
        <GlobalStyle />
        <LanguageProvider messages={translationMessages}>
          <React.Fragment>{ui}</React.Fragment>
        </LanguageProvider>
      </React.Fragment>
    </ThemeProvider>
  </Provider>
), opts)

describe('Language Select', function () {
  it('should render snapshot', function () {
    const { container } = redux(
      <LanguageSelect />
    )

    expect(container.firstChild).toMatchSnapshot()
  });

  it('should have default locale value selected', function () {
    const { getByText } = redux(
      <LanguageSelect />
    )

    // This should allow for it to work in CI, I hope
    const element = getByText(translationMessages[DEFAULT_LOCALE]["boilerplate.containers.locale.en"])
    expect(element.selected).toBeTruthy()
  });

  it('should allow the user to change locale', async function () {
    user.setup()

    const { getByRole, container } = redux(
      <LanguageSelect />
    )

    // Wait for it to change, should look at getting this data from the translations themselves
    await user.selectOptions(
      getByRole('combobox'),
      getByRole('option', {name: 'French'})
    )

    expect(getByRole('option', { name: 'Fran√ßais' }).selected).toBeTruthy()
    expect(getByRole('option', { name: 'Anglaise' }).selected).toBeFalsy()

    // Should have the options in French
    expect(container.firstChild).toMatchSnapshot()
  });
});
