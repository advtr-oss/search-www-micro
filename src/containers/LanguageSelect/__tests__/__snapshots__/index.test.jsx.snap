// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Language Select should allow the user to change locale 1`] = `
.c1 {
  background-color: #f6f8fa;
  border-radius: 4px;
  color: #24292f;
  border-style: none;
  padding: 4px 8px;
}

.c0 {
  padding: 2px;
}

<div
  class="c0"
>
  <select
    class="c1"
  >
    <option
      value="en"
    >
      Anglaise
    </option>
    <option
      value="fr"
    >
      Fran√ßais
    </option>
  </select>
</div>
`;

exports[`Language Select should render snapshot 1`] = `
.c1 {
  background-color: #f6f8fa;
  border-radius: 4px;
  color: #24292f;
  border-style: none;
  padding: 4px 8px;
}

.c0 {
  padding: 2px;
}

<div
  class="c0"
>
  <select
    class="c1"
  >
    <option
      value="en"
    >
      English
    </option>
    <option
      value="fr"
    >
      French
    </option>
  </select>
</div>
`;
